registry/spec.md

```diff
@@ -1491,71 +1632,1350 @@ An attempt to delete all versions MUST generate an error.
+text files.
+
+When using the file-based model, files with the extension `.cereg` use JSON
+encoding. Files with the extension `.cereg.yaml` or `.cereg.yml` use YAML
+encoding. The formal JSON schema for the file format is defined in the
+[CloudEvents Registry Document Schema](#cloudevents-registry-document-schema),
+which implements the Registry format and the CloudEvents Registry format.
+
+The media-type for the file format is `application/cloudevents-registry+json` for the
+JSON encoding and `application/cloudevents-registry+yaml` for the YAML encoding.
+
+The JSON schema identifier is `https://cloudevents.io/schemas/registry` and the
+`specversion` property indicates the version of this specification that the
+elements of the file conform to.
+
+A CloudEvents Registry file MUST contain a single JSON object or YAML document.
+The object declares the roots of the three sub-registries, which are either
+embedded or referenced. Any of the three sub-registries MAY be omitted.
+
+``` meta
+{
+   "$schema": "https://cloudevents.io/schemas/registry",
+   "specversion": "0.4-wip",
+   "endpoints": { ... } | "endpointsUrl": "URL",
+   "definitiongroups": { ... } | "definitiongroupsUrl": "URL"
+   "schemagroups" : { ... } | "schemagroupsUrl": "URL",
+}
+```
+
+While the file structure leads with endpoints followed by definition groups and
+then schema groups by convention, the order of the sub-registries is not significant.
+
+### Schema Registry
+
+The schema registry is a metadata store for organizing data schemas of any kind.
+
+The Registry API extension model of the Schema Registry is:
+
+``` JSON
+{
+  "groups": [
+    {
+      "singular": "schemagroup",
+      "plural": "schemagroups",
+      "schema": "TBD",
+      "resources": [
+        {
+          "singular": "schema",
+          "plural": "schemas",
+          "versions": -1,
+        }
+      ]
+    }
+  ]
+}
+```
+
+#### Group: schemagroups
+
+The group (GROUP) name for the Schema Registry is `schemagroups`. The group does
+not have any specific extension attributes.
+
+A schema group is a collection of schemas that are related to each other in some
+application-defined way. A schema group does not impose an restrictions on the
+contained schenmas, meaning that a schema group can contain schemas of different
+formats. Every schema MUST reside inside a schema group.
+
+Example:
+
+``` meta
+{
+  "schemagroups": {
+    "com.example.schemas": {
+      "type": "schemagroup",
+      "id": "com.example.schemas",
+      "schemas": {
+        ...
+      }
+    }
+  }
+}
+```
+
+#### Resource: schemas
+
+The resources (RESOURCE) collection inside of schema groups is named `schemas`.
+The type of the resource is `schema`. Any single `schema` is a container for
+one or more `versions`, which hold the concrete schema documents or schema
+document references.
+
+Any new schema version that is added to a schema definition MUST be backwards
```

Comment:
I _believe_ the text here is accurate, however I think we may need to add some text about versioning of resources someplace because this text will confuse people into thinking that they can't have a v1 and v2 of a schema doc unless they're backwards compatible. Which isn't true - we just need to make it clear that there are two levels of versioning... similar to git commits vs git branches.  Part of me wonders if we should consider using a different term than "version" just to avoid the potential confusion. We'll also need to make it clear that branch-like versioning isn't really part of the spec as much as it's a naming convention of the ID of the resources.

Let's discuss.

Reply:
hmm maybe I'm wrong - maybe the text isn't right in that you should be allowed to have breaking changes across revisions, otherwise people would never be allowed to have major version number bumps. I may have been mixing up versions with the epoch stuff.....